File: services/cart/src/telemetry.ts
// Four-signal OpenTelemetry bootstrap for the cart service (MELT).
//
// Loaded via `node -r ./dist/telemetry.js` BEFORE the app, so grpc (server) and
// ioredis (client) are auto-instrumented. Takes cart from traces-only to the
// full four correlated signals — Metrics, Events, Logs, Traces — over OTLP/gRPC,
// matching the catalogue (Go) reference and the users (Python) retrofit
// (ARCHITECTURE.md §9 / ADR-0002). One NodeSDK wires all three providers, so they
// share NodeSDK's auto-detected resource and every record agrees on service.name
// / service.namespace / deployment.environment (from the OTEL_* env in compose),
// and in-request logs/events carry the active trace_id (the correlation contract).
//
// Node-specific maturity finding (verified in Docker against the pinned SDK):
//   * Traces + Logs + Events are mature: the Logs API captures the active
//     trace_id/span_id automatically, and a domain event is a log record with the
//     `eventName` field set (the JS equivalent of Go's Record.SetEventName).
//   * Metrics RED is emitted by hand (grpc auto-instrumentation produces spans
//     only, no server metrics) — like the Go/Python interceptor.
//   * METRIC EXEMPLARS ARE NOT AVAILABLE in OpenTelemetry-JS (sdk-metrics 2.7.1,
//     the latest): the histogram aggregator never records exemplars and the OTLP
//     serializer never writes the exemplar field — verified by reading the SDK
//     source. The metric→trace link for this service is therefore provided by the
//     LGTM bundle's Tempo metrics-generator (`send_exemplars: true`), which derives
//     exemplar-bearing span metrics from cart's traces server-side. This is the
//     documented JS-stack divergence from the Go/Python reference; see §9.
import { NodeSDK } from '@opentelemetry/sdk-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-grpc';
import { OTLPLogExporter } from '@opentelemetry/exporter-logs-otlp-grpc';
import { PeriodicExportingMetricReader, AggregationType } from '@opentelemetry/sdk-metrics';
import { BatchLogRecordProcessor } from '@opentelemetry/sdk-logs';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { metrics, type Attributes, type Counter, type Histogram } from '@opentelemetry/api';
import { logs, SeverityNumber } from '@opentelemetry/api-logs';
import * as grpc from '@grpc/grpc-js';
import type { ServerInterceptor } from '@grpc/grpc-js';

// Instrumentation scope shared by the cart's own meter, logger and events —
// distinct from the grpc/ioredis auto-instrumentation scopes. Lets queries pick
// out the cart service's own metrics and domain events (in Loki, events carry
// scope_name="shoeshop/cart").
const SCOPE = 'shoeshop/cart';

// Second-scale buckets for the RPC latency histogram (the JS default buckets
// assume milliseconds and bunch up sub-second values). Mirrors the Go/Python
// reference buckets so the histogram is comparable across stacks.
const LATENCY_BUCKETS = [0.001, 0.0025, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5];

const sdk = new NodeSDK({
  // ── Traces ── unchanged from the existing wiring: one SERVER span per RPC,
  // child ioredis CLIENT spans (gRPC RPC → redis command).
  traceExporter: new OTLPTraceExporter(),
  // ── Metrics ── periodic OTLP export; the View gives rpc.server.duration
  // second-scale buckets (data-style aggregation, sdk-metrics 2.x API).
  metricReaders: [new PeriodicExportingMetricReader({ exporter: new OTLPMetricExporter() })],
  views: [
    {
      instrumentName: 'rpc.server.duration',
      aggregation: {
        type: AggregationType.EXPLICIT_BUCKET_HISTOGRAM,
        options: { boundaries: LATENCY_BUCKETS },
      },
    },
  ],
  // ── Logs + Events ── both flow through one LoggerProvider → OTLP → Loki.
  logRecordProcessors: [new BatchLogRecordProcessor(new OTLPLogExporter())],
  instrumentations: [
    getNodeAutoInstrumentations({
      '@opentelemetry/instrumentation-fs': { enabled: false },
      '@opentelemetry/instrumentation-http': {
        ignoreIncomingRequestHook: (req) => (req.url ?? '') === '/healthz',
      },
    }),
  ],
});

sdk.start();

// Meter, logger and instruments are created AFTER start(): only then has NodeSDK
// installed the global meter/logger providers (before start they are no-ops).
const meter = metrics.getMeter(SCOPE);
const otelLogger = logs.getLogger(SCOPE);

const requests: Counter = meter.createCounter('rpc.server.requests', {
  description: 'Count of inbound gRPC unary calls by method and gRPC status code.',
});
const duration: Histogram = meter.createHistogram('rpc.server.duration', {
  unit: 's',
  description: 'Duration of inbound gRPC unary calls, in seconds.',
});

/**
 * RED metrics interceptor. The gRPC auto-instrumentation emits spans only, so —
 * like the Go and Python references — the cart records its own RED:
 *   rpc.server.requests  (counter)            — rate + errors
 *   rpc.server.duration  (histogram, seconds) — duration
 * keyed by rpc.method (full path "/cart.v1.CartService/GetCart") and
 * rpc.grpc.status_code. The transport-level interceptor's sendStatus responder
 * fires once the final status is sent, so timing spans the whole call and the
 * status code is the one the client receives.
 */
export const metricsInterceptor: ServerInterceptor = (methodDescriptor, call) => {
  const startNs = process.hrtime.bigint();
  const responder = new grpc.ResponderBuilder()
    .withSendStatus((status, next) => {
      const seconds = Number(process.hrtime.bigint() - startNs) / 1e9;
      const code = status.code ?? grpc.status.UNKNOWN;
      const attributes: Attributes = {
        'rpc.method': methodDescriptor.path,
        'rpc.grpc.status_code': grpc.status[code] ?? String(code),
      };
      duration.record(seconds, attributes);
      requests.add(1, attributes);
      next(status);
    })
    .build();
  return new grpc.ServerInterceptingCall(call, responder);
};

/**
 * Structured logger that fans out to BOTH stdout JSON (so `task logs -- cart`
 * still works) AND OTLP logs → Loki. When called inside a request's span the
 * Logs API stamps the active trace_id/span_id — the trace↔log join key from the
 * correlation contract. Startup logs (outside any span) carry no trace_id, as
 * expected.
 */
function emit(severityNumber: SeverityNumber, level: string, msg: string, attributes?: Attributes): void {
  console.log(JSON.stringify({ level, msg, ...attributes }));
  otelLogger.emit({ severityNumber, severityText: level.toUpperCase(), body: msg, attributes });
}

export const log = {
  info: (msg: string, attributes?: Attributes) => emit(SeverityNumber.INFO, 'info', msg, attributes),
  warn: (msg: string, attributes?: Attributes) => emit(SeverityNumber.WARN, 'warn', msg, attributes),
  error: (msg: string, attributes?: Attributes) => emit(SeverityNumber.ERROR, 'error', msg, attributes),
};

/**
 * Emit a domain/lifecycle event via the Logs API — a log record carrying an
 * eventName (e.g. "cart.item.added"). Emitting within a request's span makes the
 * SDK stamp the active trace_id/span_id, so events join to their trace exactly
 * like logs do (the four-signal "E"). In Loki, events are told apart from logs by
 * scope_name="shoeshop/cart" + the eventName body. Richer domain events arrive
 * with the v0.3 NATS write path.
 */
export function event(name: string, attributes?: Attributes): void {
  otelLogger.emit({
    eventName: name,
    severityNumber: SeverityNumber.INFO,
    body: name,
    attributes,
  });
}

function shutdown(): void {
  sdk
    .shutdown()
    .then(() => process.exit(0))
    .catch(() => process.exit(1));
}

process.on('SIGTERM', shutdown);
process.on('SIGINT', shutdown);

File: services/users/app/telemetry.py
"""Four-signal OpenTelemetry bootstrap for the users service (MELT).

Takes users from traces-only to the full four correlated signals — Metrics,
Events, Logs, Traces — over OTLP/gRPC, matching the catalogue reference
(ARCHITECTURE.md §9 / ADR-0002). Every signal shares one resource, so all records
agree on service.name / service.namespace / deployment.environment, and
in-request metrics/logs/events carry the active trace_id (correlation contract).

All SDK APIs were verified in Docker against the pinned versions before use. Notes
on Python-specific maturity (vs the Go reference):
  * gRPC instrumentation emits spans only (no server metrics), so RED metrics come
    from a small aio interceptor here — like the Go interceptor.
  * Logs use the stdlib-logging bridge (LoggingHandler), which captures the active
    trace_id/span_id automatically.
  * The dedicated Events API is deprecated (since 1.39.0); the forward path — and
    the one used here — is a log record with the `event_name` field set, emitted
    via the Logs API (the Python equivalent of Go's Record.SetEventName).
"""

import logging
import time
from dataclasses import dataclass
from typing import Callable, Mapping, Optional

import grpc
from opentelemetry import metrics, trace
from opentelemetry._logs import SeverityNumber, set_logger_provider
from opentelemetry.exporter.otlp.proto.grpc._log_exporter import OTLPLogExporter
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.asyncpg import AsyncPGInstrumentor
from opentelemetry.sdk._logs import LoggerProvider, LoggingHandler
from opentelemetry.sdk._logs._internal import LogRecord
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
from opentelemetry.sdk.metrics import MeterProvider, TraceBasedExemplarFilter
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.metrics.view import ExplicitBucketHistogramAggregation, View
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

# Instrumentation scope shared by this service's meter and its events logger —
# distinct from the grpc/asyncpg auto-instrumentation scopes. Lets queries pick
# out the catalogue/users service's own metrics and domain events.
_SCOPE = "shoeshop/users"

# Second-scale buckets for the RPC latency histogram (the Python default buckets
# assume milliseconds and bunch up sub-second values).
_LATENCY_BUCKETS = [0.001, 0.0025, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5]

# Set in setup(); used by event() to emit domain events through the Logs API.
_event_logger = None


@dataclass
class Telemetry:
    """Wired four-signal stack handed back to main()."""

    # A gRPC aio interceptor recording RED metrics (with trace_id exemplars).
    metrics_interceptor: grpc.aio.ServerInterceptor
    # Attach to the root logger so stdlib logs are bridged to OTLP/Loki with
    # trace_id/span_id stamped on in-request records.
    logging_handler: logging.Handler
    # Flushes traces, metrics and logs on shutdown.
    shutdown: Callable[[], None]


class _MetricsInterceptor(grpc.aio.ServerInterceptor):
    """Records RED metrics per unary RPC.

    Rate/errors come from rpc.server.requests (counter by method + gRPC status);
    duration from rpc.server.duration (histogram, seconds). Recording happens
    while the tracing interceptor's server span is active, so the histogram
    carries a trace_id exemplar. Mirrors the catalogue Go interceptor.
    """

    def __init__(self, duration, requests) -> None:
        self._duration = duration
        self._requests = requests

    async def intercept_service(self, continuation, handler_call_details):
        handler = await continuation(handler_call_details)
        if handler is None or not handler.unary_unary:
            return handler

        method = handler_call_details.method
        inner = handler.unary_unary

        async def wrapper(request, context):
            start = time.perf_counter()
            try:
                return await inner(request, context)
            finally:
                code = context.code()
                status = code.name if code is not None else "OK"
                attrs = {"rpc.method": method, "rpc.grpc.status_code": status}
                self._duration.record(time.perf_counter() - start, attrs)
                self._requests.add(1, attrs)

        # The handler constructor lives on the top-level grpc module (not
        # grpc.aio); an async behavior is valid for an aio server.
        return grpc.unary_unary_rpc_method_handler(
            wrapper,
            request_deserializer=handler.request_deserializer,
            response_serializer=handler.response_serializer,
        )


def setup(service_name: str) -> Telemetry:
    # Resource.create merges OTEL_RESOURCE_ATTRIBUTES / OTEL_SERVICE_NAME from the
    # environment; the explicit service.name is a belt-and-braces default.
    resource = Resource.create({"service.name": service_name})

    # ── Traces ──────────────────────────────────────────────────────
    tracer_provider = TracerProvider(resource=resource)
    tracer_provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter()))
    trace.set_tracer_provider(tracer_provider)
    # Must run before the asyncpg pool is created so connections are patched.
    AsyncPGInstrumentor().instrument()

    # ── Metrics ─────────────────────────────────────────────────────
    duration_view = View(
        instrument_name="rpc.server.duration",
        aggregation=ExplicitBucketHistogramAggregation(boundaries=_LATENCY_BUCKETS),
    )
    meter_provider = MeterProvider(
        resource=resource,
        metric_readers=[PeriodicExportingMetricReader(OTLPMetricExporter())],
        views=[duration_view],
        exemplar_filter=TraceBasedExemplarFilter(),
    )
    metrics.set_meter_provider(meter_provider)
    meter = meter_provider.get_meter(_SCOPE)
    duration = meter.create_histogram(
        "rpc.server.duration",
        unit="s",
        description="Duration of inbound gRPC unary calls, in seconds.",
    )
    requests = meter.create_counter(
        "rpc.server.requests",
        description="Count of inbound gRPC unary calls by method and gRPC status code.",
    )

    # ── Logs + Events ───────────────────────────────────────────────
    logger_provider = LoggerProvider(resource=resource)
    logger_provider.add_log_record_processor(
        BatchLogRecordProcessor(OTLPLogExporter())
    )
    set_logger_provider(logger_provider)

    global _event_logger
    _event_logger = logger_provider.get_logger(_SCOPE)

    logging_handler = LoggingHandler(
        level=logging.INFO, logger_provider=logger_provider
    )

    def shutdown() -> None:
        tracer_provider.shutdown()
        meter_provider.shutdown()
        logger_provider.shutdown()

    return Telemetry(
        metrics_interceptor=_MetricsInterceptor(duration, requests),
        logging_handler=logging_handler,
        shutdown=shutdown,
    )


def event(name: str, attributes: Optional[Mapping[str, object]] = None) -> None:
    """Emit a domain/lifecycle event via the Logs API.

    Sets the OTLP log record's event_name field (e.g. "users.user.viewed") and
    emits it through the Logs API logger. Emitting inside a request's span makes
    the SDK stamp the active trace_id/span_id, so events join to their trace just
    like logs do. The four-signal "E"; richer domain events arrive with the v0.3
    NATS write path.
    """
    if _event_logger is None:  # setup() not called — no-op rather than crash.
        return
    _event_logger.emit(
        LogRecord(
            event_name=name,
            body=name,
            severity_number=SeverityNumber.INFO,
            attributes=dict(attributes or {}),
        )
    )
